### ISSUE TITLE
createCustomerProfile: null cardCode results in invalid XML error

### ISSUE THREAD

**Issue by alfredaday** [AuthorizeNet/sdk-java#50]
Hi,

I'm using the CIM API in your sdk.  I have a use case where I need to be able to create a customer profile with a payment profile that does not have a card code.  To do that, I set the card code to null and then call the API.  However, the SDK calls Authorize.net and receives an error message:

[Error] E00003: The 'AnetApi/xml/v1/schema/AnetApiSchema.xsd:cardCode' element is invalid - The value '' is invalid according to its datatype 'AnetApi/xml/v1/schema/AnetApiSchema.xsd:cardCode' - The Pattern constraint failed.

The reason it receives the error is because the SDK creates the cardCode element in the XML and it is left empty (Please notice the '<cardCode></cardCode>' in the request below):

```
<?xml version="1.0"?>
<createCustomerProfileRequest xmlns="AnetApi/xml/v1/schema/AnetApiSchema.xsd">
  <merchantAuthentication>
    <name>LOGIN_ID_MASKED</name>
    <transactionKey>TRANSACTION_KEY_MASKED</transactionKey>
  </merchantAuthentication>
  <profile>
    <merchantCustomerId>AAAAAA</merchantCustomerId>
    <description>df98d979-0dd0-4f98-bc66-ed654b3f1388</description>
    <email></email>
    <paymentProfiles>
      <customerType>individual</customerType>
      <billTo>
        <firstName></firstName>
        <lastName></lastName>
        <company></company>
        <address></address>
        <city></city>
        <state></state>
        <zip></zip>
        <country></country>
        <phoneNumber></phoneNumber>
        <faxNumber></faxNumber>
      </billTo>
      <payment>
        <creditCard>
          <cardNumber>MASKED**0002</cardNumber>
          <expirationDate>2029-07</expirationDate>
          <cardCode></cardCode>
        </creditCard>
      </payment>
    </paymentProfiles>
  </profile>
  <validationMode>testMode</validationMode>
</createCustomerProfileRequest>
```

Here is a test case that you can use to reproduce this issue:

```
@Test
public void testCase() {
    final String authorizeNetCardNotPresentApiLoginId = "YOUR_API_LOGIN_ID";
    final String authorizeNetCardNotPresentTransactionKey = "YOUR_TRANSACTION_KEY";

    final Merchant merchant = Merchant.createMerchant(Environment.SANDBOX, authorizeNetCardNotPresentApiLoginId, authorizeNetCardNotPresentTransactionKey);

    final CreditCard creditCard = CreditCard.createCreditCard();
    creditCard.setCreditCardNumber("370000000000002");
    creditCard.setExpirationDate("2029-07");
    creditCard.setCardCode(null);

    final Address address = Address.createAddress();
    address.setZipPostalCode(null);

    final PaymentProfile paymentProfile = PaymentProfile.createPaymentProfile();
    paymentProfile.addPayment(Payment.createPayment(creditCard));
    paymentProfile.setCustomerType(CustomerType.INDIVIDUAL);
    paymentProfile.setBillTo(address);

    final CustomerProfile customerProfile = CustomerProfile.createCustomerProfile();
    customerProfile.setMerchantCustomerId("AAAAAA");
    customerProfile.setDescription(UUID.randomUUID().toString());

    final ValidationModeType validationMode = ValidationModeType.TEST_MODE;

    // Create a new customer profile
    final net.authorize.cim.Transaction transaction = merchant.createCIMTransaction(TransactionType.CREATE_CUSTOMER_PROFILE);
    transaction.setCustomerProfile(customerProfile);
    transaction.addPaymentProfile(paymentProfile);
    transaction.setValidationMode(validationMode);

    final net.authorize.cim.Result<Transaction> result = (net.authorize.cim.Result<Transaction>) merchant.postTransaction(transaction);

    if (false == result.isOk()) {
        final StringBuilder stringBuilder = new StringBuilder();

        boolean isFirst = true;
        for (final Message message : result.getMessages()) {
            final String code = message.getCode();
            final String text = message.getText();

            if (isFirst == false) {
                stringBuilder.append(", ");
            }
            stringBuilder.append(code+": "+text);

            isFirst = false;
        }

        throw new RuntimeException("["+result.getResultCode()+"] "+stringBuilder.toString());
    }

}
```

Luckily, the fix for this is very simple.  In your net.authorize.cim.Transaction.java file, in the addPayment method you need to modify the following lines:

```
Element card_code_el = document.createElement(AuthNetField.ELEMENT_CARD_CODE.getFieldName());
card_code_el.appendChild(document.getDocument().createTextNode(credit_card.getCardCode()));
cc_el.appendChild(card_code_el);
```
All you need to do is wrap an if statement around it, like so:
```
if (StringUtils.isNotEmpty(credit_card.getCardCode())) {
    Element card_code_el = document.createElement(AuthNetField.ELEMENT_CARD_CODE.getFieldName());
    card_code_el.appendChild(document.getDocument().createTextNode(credit_card.getCardCode()));
    cc_el.appendChild(card_code_el);
}
```

**Comment by alfredaday** [comment]
Any update on this?

**Comment by brianmc** [comment]
Hi Alfred,  should have an update soon in this, thank-you for the pull request. Bin the meantime, our new model will work for this issue, check out the sample code here and let us know any feedback: https://github.com/AuthorizeNet/sample-code-java/blob/master/src/main/java/net/authorize/sample/CustomerProfiles/CreateCustomerProfile.java

Brian

**Comment by alfredaday** [comment]
Thanks for the response.  I figured that the new model wouldn't have this issue.  We have code that is still using the old model so it would be great if we could get the issue for the old model resolved :)

**Comment by brianmc** [comment]
Fixed in Release 1.8.6  http://search.maven.org/#artifactdetails%7Cnet.authorize%7Canet-java-sdk%7C1.8.6%7Cjar

Thanks for your feedback and contribution,

Brian
